### DISCUSSION TITLE
srt_connect behaviour on non-blocking mode

### DISCUSSION THREAD

**Discussion by guilletrejo** [Haivision/srt#2819] [Q&A] [answered]
Hi. So I want to setup a caller socket with an "async" connect so it never blocks. I replace the thread block with an "interruptible sleep" that I do with a condition variable. So my code looks something like this:

```
    srtSock = srt_create_socket();
    const bool connBlock = false;
    srt_setsockflag(srtSock, SRTO_RCVSYN, &connBlock, sizeof(connBlock))
    const int32_t timeout = 5000;
    srt_setsockflag(srtSock, SRTO_CONNTIMEO, &timeout, sizeof(timeout))

    auto ret = srt_connect(srtSock, dstAddr, dstAddrSize);
    while() {
        interruptibleSleep(std::chrono::milliseconds(1000));
        int32_t connectedState;
        int ssize = sizeof(connectedState);
        srt_getsockflag(srtSock, SRTO_EVENT, &connectedState, &ssize);
        if (connectedState == SRT_EPOLL_CONNECT) {
            break;
        }
    }
    ... rest of the program with the connection established
```

I would hope that if no client is listening on `dstAddr`, this code will loop forever waiting for the `EPOLL_CONNECT` event. However, via debug logs I am seeing that it actually only tries to connect for 5 seconds (the `SRTO_CONNTIMEO` that I set) and then the socket goes to a `SRTS_BROKEN` state. Is this the expected behaviour? My only workaround is to create the socket again when `SRTO_CONNTIMEO` has passed and there was no connection.

Is this how is is supposed to be used? It would be very great to be able to have an "infinite" timeout so I know the socket won't be created again and again if the program is running a lot of time with no clients.

Thanks!

**Comment by ethouris** [comment, marked as answer]
Broken sockets cannot be reused and there's no intended "infinite" value for `SRTO_CONNTIMEO`. So yes, the only solution is to recreate the socket on failure.

Not sure why you need non-blocking mode for such a scenario. You are doing the same thing as it would be with blocking mode, just with more idle calls.

**Reply by guilletrejo** [reply]
Thanks a lot for the answer! Actually I need this only because the interruptible sleep. I want to be able to call a `stop()` function from outside this thread so it stops immediately. In blocking mode, when the thread blocks on srt_connect there's no way to wake it before the timeout.
